### PR TITLE
feat: release a windowed shuffled pass's spill per block and re-read it

### LINKED ISSUE #66: Windowed + shuffled residency: the ordering, not the window, holds the whole split
Windowed residency is not caused by windows, and not by shuffle. It is caused by the two
together, and today's ordering makes the combination as bad as it can be.

## The 2x2

512-sample store, one sample per chunk, 410 train chunks, `block_chunks=8`, `batch_size=8`.
"Live set" is the peak number of chunks that must be co-resident; "lifetime" is how many
blocks a chunk stays resident between its first and last use.

| | no shuffle | shuffle |
|---|--:|--:|
| **no window** `[0]` | 16 chunks (4%) | 16 chunks (4%) |
| **window** `[0, 1]` | 18 chunks (4%) | **229 chunks (56%)** |

Same offsets, same batch size, same block size: shuffle turns an 18-chunk working set into
229. Mean chunk lifetime goes from 1.1 blocks to 18.0, with a maximum of 51 out of 52 --
some chunk is needed in the first block of the epoch and again in the last.

That is why `working_set_bytes` clamps a windowed shuffled pass to the whole train split.
The clamp is roughly honest: measured live sets are 46-56% of the split against a clamp of
100%.

## Why

`block_shuffled_order` permutes at **chunk** granularity, then slices blocks off that
permutation. A windowed anchor `c` reads chunks `c` and `c+lead`, and chunk `c+lead` is
*also* the anchor chunk of another draw -- so two anchors need the same chunk, and the
permutation puts them in uniformly random blocks. Since the driver fetches each chunk once
per epoch (decode-once), that chunk must stay resident across the gap between them.

Residency is therefore set by chunk **lifetime** under the permutation, not by how many
chunks a batch touches. It gets worse as leads multiply, because each additional lead adds
another pair of anchors that must both be alive:

| offsets | live set | % of split | mean lifetime |
|---|--:|--:|--:|
| `[0]` | 16 | 4% | 1.0 |
| `[0, 24]` | 223 | 54% | 16.7 |
| `[0, 24, 240]` | 190 | 46% | 7.7 |
| `[0, 24, 48, 72]` | **397** | **97%** | 27.1 |

A 24/48/72-hour multi-lead training set -- an ordinary forecasting setup -- holds
essentially the entire split resident.

## Exploratory: ordering along the lead chain

The anchors that share a chunk are not random. For a lead stride `s` they are exactly
`c, c+s, c+2s, ...`. Emitting a chain together makes those anchors adjacent, so the chunk's
lifetime collapses -- and because a chain is spread across the whole store, consecutive
emitted samples are still far apart in source order, unlike a contiguous-run ordering.

| offsets | order | live set | mean lifetime | `shuffle_quality` |
|---|---|--:|--:|--:|
| `[0, 24]` | chunk perm (today) | 223 | 16.7 | 0.99 |
| `[0, 24]` | chain perm | **19** | 1.1 | 0.32 |
| `[0, 24, 48, 72]` | chunk perm (today) | 397 | 27.1 | 0.99 |
| `[0, 24, 48, 72]` | chain perm | **25** | 1.3 | 0.32 |

9-16x less residency. What it costs is mixing, and possibly more than the metric says: the
samples a chain groups are exactly one lead apart, which on a weather store is the same
time of day, so they are diurnally correlated in a way `shuffle_quality` does not see.

A first skill check (synthetic advection store, 1536 steps, 24-step horizon, CPU): model
RMSE 0.4837 (chunk) vs 0.4899 (chain) after one epoch, against an identical persistence
RMSE of 0.7249 in both -- so the loader returns the same data, and the skill difference is
about ordering. One epoch and one seed is noise, and a repeated comparison is running.

**Contiguous-run ordering does not work** and was measured before this: it helps only when
the lead is smaller than a run (`[0,1]`: 229 -> 41) and is *worse* at a 24-hour lead
(240 vs 223), while costing the same mixing. Chain ordering follows the stride, which is
why it survives where runs do not.

## What to decide

1. Is 0.32 mixing acceptable for a windowed dataset, given the samples grouped are one
   lead apart? This is an ML question, and the persistence-RMSE harness above is the
   instrument for it.
2. If it is, does chain ordering become automatic when offsets are present (there is no
   sane reason to choose the 97%-residency ordering), or a second shuffle mode? A knob here
   would sit badly with "one obvious way to do it".
3. `[0, 1]` degenerates to sequential order (`shuffle_quality` 0.01), so any rule has to be
   conditional on the lead stride relative to `block_chunks`.

Not a change to make quietly: it alters the order samples arrive in, which is a training
contract, not an implementation detail. Hence an issue rather than a PR.

Related: #65 (bounded read-ahead and the residency floor -- the measurements above use its
`read_ahead_bound` to compute live sets), and the "re-fetch the spill" note in
`working_set_bytes`, which is the other way out of the same problem and trades decode-once
for memory instead of trading mixing.

### PR DESCRIPTION
**Pull request by emfdavid**
## Summary

Closes #66.

A windowed view reads `anchor + offset` and shuffle permutes chunk order, so a chunk one block
reads can be needed again most of an epoch later. Held from first use to last, residency was
close to the whole train split rather than two blocks.

With `persist=True` each block hands its chunks back as it drains, and a later block re-admits
them from the on-disk cache.

| offsets | policy | floor (chunks) | peak resident | re-reads | samples |
|---|---|---|---|---|---|
| `{0, 12}` | retain | 259 | 256 | 0 | 1012 |
| `{0, 12}` | **release** | **12** | **12** | 249 | 1012 |
| `{0, 24, 240}` | retain | 316 | 256 | 0 | 784 |
| `{0, 24, 240}` | **release** | **18** | **18** | 331 | 784 |

256-chunk split, `block_chunks=2`, two views of one array. Same samples delivered either way.

## Three parts, none of which works alone

* **plan** — dedup scoped to a block (`groups=`), so a chunk two blocks apart is admitted twice;
  the driver decides admission per chunk-run rather than once per pass.
* **consumer** — releases every key its block read, not only those at their last use; the
  read-ahead bound becomes three consecutive blocks (a batch on a boundary transiently holds the
  block behind it, the one it gathers, and the one it awaits).
* **floor** — `working_set_bytes` learns the policy. This one is easy to miss: a released chunk
  is only *evictable*, so with a floor still sized for retention nothing is ever evicted and
  releasing changes no number at all. That was the state of the parked spike.

## Two things measurement corrected

**The rule keys on `persist`, not `cache_dir`.** Only a persisted slot survives its own eviction
as a revivable file; with `cache_dir` alone the backing is unlinked on eviction, so the re-read
pays the fetch *and* the decode again — strictly worse than having retained the chunk. Same
fixture, same 249 re-reads:

```
persist=False   hits  10/501  (2%)   cache files   12
persist=True    hits 246/502  (49%)  cache files  256
```

The first draft of this keyed on `cache_dir` and logged "served from cache_dir", which would
have shipped a claim that was false in exactly the configuration it described.

**Why #66 was blocked, and no longer is.** The parked spike deadlocked because the driver planned
a block from its chunk ids while the consumer released the keys its rows read, and those
disagreed on 66 of 77 blocks. #68/#69 fixed that: block membership now comes from the builder, so
the planner's clamp and the consumer's edge-anchor drop agree. Verified before building on it —
across 11 geometries (plain, leads of 1/one-chunk/three-chunk, history+lead, sparse leads,
one-sample chunks, fat chunks, ragged tail, wide blocks, negative-only, a lead past the split):
**0 surplus permits, 0 shortfall, every block.**

## For reviewers

The `releases_spill` argument on `working_set_bytes` is the part worth a second look — it changes
a floor that raises at construction when a budget is below it, so getting it too low starves a
pass mid-epoch, which fails in the shape of a hang. It is three blocks rather than two for the
boundary reason above.

`docs/tuning.md` gains a section: what the trade is, why it is a rule and not a knob, and how to
read the churn line. The per-epoch summary reports re-reads and evictions whenever non-zero,
because a hit rate cannot show this — a re-read served from the cache counts as a hit.

Tests: 7 in `tests/test_reread_spill.py`, including the two controls that matter — an unwindowed
pass must not enable the policy (it would buy nothing and cost the dedup), and a retained pass
must report no re-reads (if it ever does, either the plan stopped deduplicating or a chunk is
being admitted twice for no reason).

## Author attestation

- [ ] I have reviewed every change in this PR, I can explain why each one is correct, and I
      have verified the claims made in this description.

Left unticked deliberately: it asserts that a human reviewed every change and verified the
claims above, which is yours to assert and not mine to assert on your behalf.

## Checklist

- [x] Tests added or updated
- [x] `ruff check`, `ruff format --check`, `mypy`, `pytest -q` (651 passed + tf's 2 in its own
      process) and `mkdocs build --strict` green locally
- [x] Docstrings for the changed surface (`working_set_bytes`, `Scheduler.start`,
      `build_stored_chunk_reads`)
- [x] User-facing behaviour documented in `docs/tuning.md`
- [x] A bullet added under `## Unreleased` in `CHANGELOG.md`
- [x] No load-bearing invariant is broken
- [ ] Touches `ChunkPool` / the scheduler / cross-thread readiness → 3.13t run passes (CI)

### CODE REVIEW

File: src/insitubatch/source.py [hunk 1]
```diff
@@ -259,6 +259,20 @@ def __init__(
         # The last completed pass, or None before one finishes. Per pass, not per epoch: a
         # training epoch iterates train then val over one pool, and averaging two passes
         # with different shapes into one report is how a number stops meaning anything.
+        # A windowed shuffled pass holds a chunk from its first use to its last, which under
+        # a chunk permutation is most of an epoch -- so residency is close to the whole split
+        # (#66). Releasing each block's chunks when that block drains returns residency to the
```
Comment (emfdavid) [inline src/insitubatch/source.py]: Remove retrospective view

**Review by emfdavid — COMMENTED** [review]
:shipit:

File: src/insitubatch/source.py (hunk 1 above)
Comment (emfdavid) [reply src/insitubatch/source.py]: Dropped, and the same framing in three neighbours I wrote: `ChunkPool._note_touch`, the admission-decision comment in `_drive`, and `working_set_bytes`. Each explained the design by contrast with what it replaces, and the `reread_spill` one also carried the measurement that chose between the two.

All four now state the current rule. The measurement (249 re-reads served 2% from cache without `persist`, 49% with it) is in the CHANGELOG entry, which is where provenance belongs.

Swept the rest of the diff for the same pattern — three remaining hits on "would" are forward-looking statements about current behaviour, not retrospect.

652 passed, mypy and ruff clean.
